[pkg] Catch exceptions where dependency resolution gets invalid imports

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58573 [pkg] Catch exceptions where dependency resolution gets invalid imports**
* #58572 [pkg] simplifications to broken dependency handling

Users can create invalid imports, like:
```
HG: in a top-level package
if False:
  from .. import foo
```

Since this code is never executed, it will not cause the module to fail to
load. But our dependency analysis walks every `import` statement in the AST,
and will attempt to resolve the (incorrectly formed) import, throwing an exception.

For posterity, the code that triggered this: https://git.io/JsCgM

Differential Revision: [D28543980](https://our.internmc.facebook.com/intern/diff/D28543980/)

Differential Revision: [D28543980](https://our.internmc.facebook.com/intern/diff/D28543980)

suo committed:
[pkg] Catch exceptions where dependency resolution gets invalid imports

Users can create invalid imports, like:
```
HG: in a top-level package
if False:
  from .. import foo
```

Since this code is never executed, it will not cause the module to fail to
load. But our dependency analysis walks every `import` statement in the AST,
and will attempt to resolve the (incorrectly formed) import, throwing an exception.

For posterity, the code that triggered this: https://git.io/JsCgM

Differential Revision: [D28543980](https://our.internmc.facebook.com/intern/diff/D28543980/)

[ghstack-poisoned]

suo committed:
Update on "[pkg] Catch exceptions where dependency resolution gets invalid imports"


Users can create invalid imports, like:
```
HG: in a top-level package
if False:
  from .. import foo
```

Since this code is never executed, it will not cause the module to fail to
load. But our dependency analysis walks every `import` statement in the AST,
and will attempt to resolve the (incorrectly formed) import, throwing an exception.

For posterity, the code that triggered this: https://git.io/JsCgM

Differential Revision: [D28543980](https://our.internmc.facebook.com/intern/diff/D28543980/)

Differential Revision: [D28543980](https://our.internmc.facebook.com/intern/diff/D28543980)

[ghstack-poisoned]

SplitInfinity commented on torch/package/package_exporter.py:
```diff
@@ -235,7 +235,16 @@ def _get_dependencies(
+        except Exception as e:
+            self.dependency_graph.add_node(
+                module_name,
+                error=PackagingErrorReason.DEPENDENCY_RESOLUTION_FAILED,
+                error_context=str(e),
```
Since this will eventually become a `PackagingError`, would it be useful to [chain `e` to the eventual `PackagingError` that will be thrown](https://docs.python.org/3/tutorial/errors.html#exception-chaining)?

Since the `PackageError` isn't raised right away, we'd have to store `e` to use in a `raise...from` later. I don't know if that's advised. :S

suo replied:
The primary problem is that PackagingError may have a huge number of causes, so it's hard to raise from just one.

SplitInfinity replied:
That's true sigh

SplitInfinity commented on test/package/test_dependency_api.py:
```diff
@@ -265,6 +265,14 @@ def import_module(self, module_name):
+        buffer = BytesIO()
+        with self.assertRaisesRegex(PackagingError, "attempted relative import"):
+            with PackageExporter(buffer, verbose=False) as exporter:
+                # This import will fail to load
```
```suggestion
                # This import will fail to load.
```

SplitInfinity commented on test/package/test_dependency_api.py:
```diff
@@ -265,6 +265,14 @@ def import_module(self, module_name):
+    def test_invalid_import(self):
+        """An incorrectly-formed import should raise a PackagingError."""
+        buffer = BytesIO()
+        with self.assertRaisesRegex(PackagingError, "attempted relative import"):
```
Consider checking that the exception message also contains the right string for `DEPENDENCY_RESOLUTION_FAILED` (i.e. `Dependency resolution failed`).

SplitInfinity reviewed (approved):
📦

suo committed:
Update on "[pkg] Catch exceptions where dependency resolution gets invalid imports"


Users can create invalid imports, like:
```
HG: in a top-level package
if False:
  from .. import foo
```

Since this code is never executed, it will not cause the module to fail to
load. But our dependency analysis walks every `import` statement in the AST,
and will attempt to resolve the (incorrectly formed) import, throwing an exception.

For posterity, the code that triggered this: https://git.io/JsCgM

Differential Revision: [D28543980](https://our.internmc.facebook.com/intern/diff/D28543980/)

Differential Revision: [D28543980](https://our.internmc.facebook.com/intern/diff/D28543980)

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28543980).
